tests/core/test_serializable.py

```diff
@@ -0,0 +1,484 @@
+        (slice(None, 2), (5, b'a')),
+        (slice(None, 1), (5,)),
+        (slice(None, 0), tuple()),
+        (slice(0, None), (5, b'a', (0, 1))),
+        (slice(1, None), (b'a', (0, 1))),
+        (slice(2, None), ((0, 1),)),
+        (slice(2, None), ((0, 1),)),
+        (slice(None, None), (5, b'a', (0, 1))),
+        (slice(-1, None), ((0, 1),)),
+        (slice(-2, None), (b'a', (0, 1),)),
+        (slice(-3, None), (5, b'a', (0, 1),)),
+    ),
+)
+def test_serializable_getitem_lookups(type_1_a, lookup, expected):
+    actual = type_1_a[lookup]
+    assert actual == expected
+
+
+def test_serializable_subclass_retains_field_info_from_parent():
+    obj = RLPType4(2, 1, 3)
+    assert obj.field1 == 1
+    assert obj.field2 == 2
+    assert obj.field3 == 3
+
+
+def test_deserialization_for_custom_init_method():
+    type_3 = RLPType3(2, 1, 3)
+    assert type_3.field1 == 1
+    assert type_3.field2 == 2
+    assert type_3.field3 == 3
+
+    result = decode(encode(type_3), sedes=RLPType3)
+    assert result.field1 == 1
+    assert result.field2 == 2
+    assert result.field3 == 3
+
+    result_sedes_encode = decode(encode(type_3, RLPType3), sedes=RLPType3)
+    assert result_sedes_encode.field1 == 1
+    assert result_sedes_encode.field2 == 2
+    assert result_sedes_encode.field3 == 3
+
+
+def test_serializable_sedes_inference(type_1_a, type_1_b, type_2):
+    assert infer_sedes(type_1_a) == RLPType1
+    assert infer_sedes(type_1_b) == RLPType1
+    assert infer_sedes(type_2) == RLPType2
+
+
+def test_serializable_invalid_serialization_value(type_1_a, type_1_b, type_2):
+    with pytest.raises(SerializationError):
+        RLPType1.serialize(type_2)
+    with pytest.raises(SerializationError):
+        RLPType2.serialize(type_1_a)
+    with pytest.raises(SerializationError):
+        RLPType2.serialize(type_1_b)
+
+
+@pytest.mark.parametrize(
+    'value,sedes',
+    (
+        # Less than 4 bytes of serialized data
+        (b'\x00\x00\x00', RLPType1),
+        (b'\x00\x00\x00', RLPType2),
+        (b'\x00\x00\x00', RLPType3),
+        (b'\x00\x00\x00', RLPType4),
+
+        # Insufficient serialized container data as per found out container length
+        (b'\x00\x00\x00\x04', RLPType1),
+        (b'\x00\x00\x00\x04', RLPType2),
+        (b'\x00\x00\x00\x04', RLPType3),
+        (b'\x00\x00\x00\x04', RLPType4),
+
+        # Serialized data given is more than what is required
+        (b'\x00\x00\x00\x08\x00\x00\x00\x00\x00\x00\x00\x01' + b'\x00', RLPType1),
+        (b'\x00\x00\x00\x08\x00\x00\x00\x00\x00\x00\x00\x01' + b'\x00', RLPType2),
+        (b'\x00\x00\x00\x08\x00\x00\x00\x00\x00\x00\x00\x01' + b'\x00', RLPType3),
+        (b'\x00\x00\x00\x08\x00\x00\x00\x00\x00\x00\x00\x01' + b'\x00', RLPType4),
+    ),
+)
+def test_container_deserialize_bad_values(value, sedes):
+    with pytest.raises(DeserializationError):
+        sedes.deserialize(value)
+
+
+def test_serializable_serialization(type_1_a, type_1_b, type_2):
```

Comment:
1. I think you can use `@pytest.mark.parametrize`?
2. Where are these result came from?

Reply:
@hwwhww yes, I will parametrize them. The results were generated using `SSZ` and not `RLP`. I took extra care there.